tapiriik/services/Aerobia/aerobia.py

```diff
@@ -0,0 +1,340 @@
+            username = CredentialStore.Decrypt(record.ExtendedAuthorization["Email"])
+
+        self._patch_user_agent()
+        request_parameters = {"user[email]": username, "user[password]": password}
+        res = requests.post(self._loginUrlRoot, data=request_parameters)
+
+        if res.status_code != 200:
+            raise APIException("Login exception {} - {}".format(res.status_code, res.text), user_exception=UserException(UserExceptionType.Authorization))
+
+        res_xml = etree.fromstring(res.text.encode('utf-8'))
+
+        info = res_xml.find("info")
+        if info.get("status") != "ok":
+            raise APIException(info.get("description"), user_exception=UserException(UserExceptionType.Authorization))
+
+        user_id = int(res_xml.find("user/id").get("value"))
+        user_token = res_xml.find("user/authentication_token").get("value")
+
+        return user_id, user_token
+
+    def _call(self, serviceRecord, request_call, *args):
+        retry_count = 3
+        resp = None
+        for i in range(0, retry_count):
+            try:
+                resp = request_call(args)
+                break
+            except APIException:
+                # try to refresh token first
+                self._refresh_token(serviceRecord)
+            except requests.exceptions.ConnectTimeout:
+                # Aerobia sometimes answer like
+                # Failed to establish a new connection: [WinError 10060] may happen while listing.
+                # wait a bit and retry
+                time.sleep(.2)
+        if resp is None:
+            raise APIException("Api call not succeed", user_exception=UserException(UserExceptionType.DownloadError))
+        return resp
+
+    def _refresh_token(self, record):
+        logger.info("refreshing auth token")
+        user_id, user_token = self._get_auth_data(record=record)
+        record.Authorization.update({"OAuthToken": user_token})
+
+    def _with_auth(self, record, params={}):
+        params.update({"authentication_token": record.Authorization["OAuthToken"]})
+        return params
+
+    def Authorize(self, username, password):
+        from tapiriik.auth.credential_storage import CredentialStore
+        user_id, user_token = self._get_auth_data(username=username, password=password)
+
+        secret = {
+            "Email": CredentialStore.Encrypt(username), 
+            "Password": CredentialStore.Encrypt(password)
+            }
+        authorizationData = {"OAuthToken": user_token}
+        
+        return (user_id, authorizationData, secret)
+
+    def DownloadActivityList(self, serviceRecord, exhaustive=False):
+        self._patch_user_agent()
+
+        activities = []
+        exclusions = []
+
+        fetch_dairy = lambda page=1: self._get_dairy_xml(serviceRecord, page)
+        # use first query responce to detect pagination options as well
+        dairy_xml = self._call(serviceRecord, fetch_dairy)
```

Comment:
Rather than having two places calling this, it would be better to do it all in the main loop (don't really need to know the total page count beforehand - just have the page counter start at 1, incrementing each loop, and break from the loop once there are no more activities are returned.